### DISCUSSION TITLE
Authentification issue

### DISCUSSION THREAD

**Discussion by alexeyaseev** [apache/airflow#24025] [General]
Hello, 

Thank you for taking your time, we have encountered auth issue when migrated from 2.1.4 to 2.3.0. We used /api/v1/security/login to get access token and then trigger dags, but after migration something looks broken.

Http://webserver:8080/api/v1/security/login gives 404 error, while on previous versions of airflow it worked (without additional params it produced 405 error, but now it produce 404 error)

It could be easily reproduced with docker compose

**Comment by potiuk** [comment]
You should look at FAB https://flask-appbuilder.readthedocs.io/en/latest/breaking.html release notes. You are likely hitting some of the breaking changes introduced in one of the FAB releases (FAB is used for authentication mechanism and the configuration you have is really FAB authtentication).

**Reply by alexeyaseev** [reply]
We actually do not use any config, we just use official docker image and docker compose (https://airflow.apache.org/docs/apache-airflow/stable/start/docker.html)
(to get docker compose we use: curl -LfO 'https://airflow.apache.org/docs/apache-airflow/2.3.2/docker-compose.yaml')

**Reply by potiuk** [reply]
If you need some help - please post the exact reproduction steps (and commands and logs that you see produced while trying) detailing exactly what to do how you authenticate, what kind of system you have etc. What do the logs of the webserver show you ? how exactly you run airflow with the compose, how did it look beofore.

Maybe taking a look at those will even help you to diagnose your problem.

Note that quickstart docker-compose is just that - quickstart. It's not fit for production and it does not have any guarantees that what worked before will continue working the same way after new version is released, so when there are any problems with it you cannot expect that this will "always work the same" - and if you have problems with it, you need to provide details on what to do - because this is just quistart/development tool.

**Reply by potiuk** [reply]
Idealy look at the differences you saw in the logs in the old version with the new one and point them out.

**Reply by potiuk** [reply]
> Even if many users think of Docker Compose as “ready to use”, it is really a developer tool, that requires the user to know very well how docker images, containers, docker compose networking, volumes, naming, image building works. It is extremely easy to make mistakes that lead to difficult to diagnose problems and if you are not ready to spend your own time on learning and diagnosing and resolving those problems on your own do not follow this path. You have been warned.

**Reply by potiuk** [reply]
This is quote from our docs.

**Reply by aserechenko** [reply]
Well, facing the same issue.
Step to reproduce:
1. Install `pip install airflow==2.3.0`
2. `airflow users create -r Admin -u admin -e a@e.com -f admin -l admin -p admin` (to prevent some errors without users)
3. Start `airflow webserver`
4. `curl -XPOST http://localhost:47824/api/v1/security/login -d '{"username": "admin", "password": "admin", "provider": "db"}' -H "Content-Type: application/json"`

Curl will return this:
```
{
  "detail": "The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.",
  "status": 404,
  "title": "Not Found",
  "type": "about:blank"
}
```

which is exactly, what isn't expected.

**Reply by aserechenko** [reply]
> You should look at FAB https://flask-appbuilder.readthedocs.io/en/latest/breaking.html release notes.

setup.cfg for Airflow==2.3.0 is using `flask-appbuilder==3.4.5`, which is fits requirements for 2.1.5: `flask-appbuilder>=3.3.2, <4.0.0`

**Reply by potiuk** [reply]
The configuration for authentiucation is done by FAB and it has changed in FAB 3.4.5 (apparently slightly). Please take a look at FAB documentation about this @aserechenko. We are merely passing the configuration to FAB.

**Reply by potiuk** [reply]
You can also add your question at FAB github repo - specify the FAB version before and after, and your configuration there, and I am quite sure they are the best to help in your problem.

**Reply by potiuk** [reply]
You can even see the message about changes in FAB 3.0 in the link passed above:

https://flask-appbuilder.readthedocs.io/en/latest/breaking.html#version-3-0-0-oauth

Did you apply those changes that FAB maintainers pointed at @aserechenko  ?

**Reply by aserechenko** [reply]
> https://flask-appbuilder.readthedocs.io/en/latest/breaking.html#version-3-0-0-oauth

I'm not sure changes made in version 3.0.0 could also affect upgrade from 3.3.2 to 3.4.5. As well this link as about OAuth, which is not the case.
And I've tried downgrading FAB back to version 3.3.2, but keeping Airflow version 2.3.0 and the problem is still here.

**Reply by potiuk** [reply]
Not sure either - best to check in FAB.

**Reply by potiuk** [reply]
There are also a number of issues/discussion about oauth/FAB link discussed here  or in Stack Overflow - they might be helpful in looking for a solution.

**Reply by aserechenko** [reply]
I've investigated this error in details, and here is the problem.

Currently, airflow fully redefine `BaseSecurityManager` from the scratch here: https://github.com/apache/airflow/blob/c95081406fa47ee737fa7937be4f13e47b2befca/airflow/www/fab_security/manager.py#L116

without inheriting original FAB `BaseSecurityManager`
https://github.com/dpgaspar/Flask-AppBuilder/blob/e0e94acbfcea23866560454ce12fe7204472496d/flask_appbuilder/security/manager.py#L131

And this fully overridden `BaseSecurityManager` having a difference in method `register_views()`, which is originally [doing](https://github.com/dpgaspar/Flask-AppBuilder/blob/e0e94acbfcea23866560454ce12fe7204472496d/flask_appbuilder/security/manager.py#L695
):
```python    
def register_views(self):
        if not self.appbuilder.app.config.get("FAB_ADD_SECURITY_VIEWS", True):
            return
        # Security APIs
        self.appbuilder.add_api(self.security_api)
```

But there is no line `self.appbuilder.add_api(self.security_api)` in overridden `BaseSecurityManager`:
https://github.com/apache/airflow/blob/c95081406fa47ee737fa7937be4f13e47b2befca/airflow/www/fab_security/manager.py#L703-L707

and there is no `security_api` attribute as well.

This leads to missing mentioned URLs.

**Reply by aserechenko** [reply]
Resolved simply by adding     `security_api = SecurityApi` attribute and then:
```python
        # Security APIs
        self.appbuilder.add_view_no_menu(self.security_api)
```
in BaseSecurityManager.register_views()

**Reply by aa3pankaj** [reply]
@potiuk looks like again we have some compatibility issue between airflow's security manager and FAB's manager.
I can confirm that syncing FAB and airflow's register_views method is fixing this issue.

Airflow version: 2.4.2
FAB version: 4.1.4

**Reply by potiuk** [reply]
Hey @gmsantos can you please take a look at that and fix ? 

You did check in https://github.com/apache/airflow/pull/26393 

```
No changes that should be replicated in airflow/www/fab_security was found.

See https://github.com/dpgaspar/Flask-AppBuilder/compare/v4.1.3...v4.1.4
```

(and I double-checked 😁  that we have no security_manager issues, but I think there are other changes that also should be copied over)

@aserechenko. - can you please tell which of the changes you copied ? I guess oauth stuff?

**Reply by gmsantos** [reply]
I can try to reproduce it here with an Azure AD oauth config.

**Reply by potiuk** [reply]
Hey @gmsantos  - did you have chance to check it ? @aa3pankaj - maybe you can submit your changes if you've already done so?

**Reply by gmsantos** [reply]
@potiuk I could confirm this difference in register_views, but this change [is really old, from 2019](https://github.com/dpgaspar/Flask-AppBuilder/pull/929) (added in FAB 3.1.1).

I can investigate in the Airflow side if this change was added some point in the past or if it was removed from Airflow register_views for some reason. I can see in this same method other differences referent to some labels (see https://github.com/dpgaspar/Flask-AppBuilder/blame/e0e94acbfcea23866560454ce12fe7204472496d/flask_appbuilder/security/manager.py#L776-L800)

**Reply by gmsantos** [reply]
so, here is where this file was added https://github.com/apache/airflow/commit/48d4c5da19217174c8996b2882bb71f40381ae2c. The beginning of this method wasn't changed since then.

I will review the whole file looking for other differences or check if is possible only to extend the FAB base class (no need to sync all changes from now on manually). 

I can try bumping to [FAB latest](https://github.com/dpgaspar/Flask-AppBuilder/releases/tag/v4.1.6) at the same time. In the meantime, feel free to submit a hotfix by adding security_api @aa3pankaj or @aserechenko 

I assuming this issue only affects API calls, right? I couldn't reproduce this issue with airflow webserver OAuth2 login flows with Azure AD.

**Reply by gmsantos** [reply]
I see why we don't extend the base FAB class https://github.com/apache/airflow/pull/21294#issuecomment-1029113190

Let me investigate a bit more and think about some mid-term change.
